### PR TITLE
Remove ineffective JVM version tracking

### PR DESCRIPTION
**Pull request by sorin-florea**
Most (or all?) types that we track are loaded via one of our classloaders, and thus we never hit the system classloader. We only use the JVM version when tracking a task, transform or nested type that was loaded directly through the system classloader.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by sorin-florea** [comment]
@bot-gradle test this

**Comment by sorin-florea** [comment]
@bot-gradle test this

**Comment by sorin-florea** [comment]
Looking at the [documentation](https://docs.gradle.org/current/userguide/common_caching_problems.html#java_version_tracking) I think I don't have to change anything in that part. I couldn't find anything else that needs to be changes.

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by lptr** [comment]
Could you update the description of the PR to explain why we are removing this?

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: That is strange. I think we should still add those two classloaders to the known classloaders, don't you think?

**Review by wolfs — CHANGES_REQUESTED** [review]

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: What is a use case where we'd need them?

Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: I don't have a use-case. Though removing the classloaders here could be a breaking change, right? Do you know why they have been here in the first place?

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: I don't have a clear memory, but here's my theory. This was added very early in the _task output caching_ project. We were very anxious about false positive cache hits, and it was unclear if there were significant differences in how JVMs worked wrt things like e.g. `HashMap` ordering or `hashCode()` implementations for standard library types etc. We wanted to err on the side of caution, but we also made a mistake, and never captured the Java version as originally intended. We also never checked (there is no test) if this piece of code actually had any effect.

As to whether it's okay to remove this... In short: I'm not 100% sure, though no test has failed, which is a good sign. For this piece of code to have any effect, you'd need to an input that is loaded via the system classloader, i.e. it has to have a type that is defined in the Java runtime library. The ones I can think of (`Integer`, `String`, `File`, `Map`) have special treatment as primitive-like types, and we never snapshot their `Class`. `Callable` might be an issue, but you never actually have a `Callable`, instead you have an implementation of `Callable` that _is_ loaded via the Gradle classloaders, so no problem there either. So all-in-all I expect that this code is never triggered and can be safely removed.

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: I don't think we have such a type, because if we did, that test should have broke. What type should we add do you think? My main problem is that I don't really have a good idea for one.

Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: Types from the `java.time` package would be good candidates I think.

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: Ah, like `java.time.Duration`. That makes sense. I think with the changes here we'd make a task having a property of type `Duration` be out-of-date all the time, since the classloader for that type is unknown after the change.

So I think a better change here would be to keep adding the classloaders, but set their hash to a constant, say, `Hashing.signature("SYSTEM")` or something.

In any case, let's add the test first, and then check if it indeed fails, and then fix it by re-adding the system classloaders with the fixed hash.

Comment (sorin-florea) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: @lptr I added a test just to check that a task having a property of type `Duration` is out-of-date all the time but it turns out it isn't

https://github.com/gradle/gradle/pull/20300/files#diff-97bc9f3d972bb9ec46c88ab98e683e8638a60555c4fec30295509a94bfd6e554R487

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: We actually capture the classloader further down the line, but `Duration` has `null` (because Java thinks it's a primitive), and we later on ignore missing classpath hashes here:

https://github.com/gradle/gradle/blob/c7f67221c5cf025430542a2bf8e3ca6dae3a61e0/subprojects/model-core/src/main/java/org/gradle/internal/snapshot/impl/JavaSerializedValueSnapshot.java#L90-L98

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: Update: it's not just for primitives. Here's `Class.getClassLoader()`'s Javadocs:

```java
    /**
     * Returns the class loader for the class.  Some implementations may use
     * null to represent the bootstrap class loader. This method will return
     * null in such implementations if this class was loaded by the bootstrap
     * class loader.
```

So the bootstrap classloader is always represented as `null`.

**Comment by sorin-florea** [comment]
@bot-gradle test this

**Comment by sorin-florea** [comment]
@lptr I incorporated your feedback. If we never took Java runtime's version into account I'm not sure what would be the implications of starting to do it now.

**Review by lptr — APPROVED** [review]
LGTM, let's wait for @wolfs's review before merging though.

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: Ok then, we have the test now so I suppose that is good enough.

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:37]: So we are having `null` as a key in the map of known classloaders now? I suppose this works though it is unexpected. Can we add a comment somewhere why we do this?

**Review by wolfs — COMMENTED** [review]
Looks good, I left a comment.

**Review by wolfs — APPROVED** [review]

**Review by wolfs — APPROVED** [review]
LGTM!

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: Is this _some_ or _all_ implementations? I think it's the latter..?

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java
Comment (sorin-florea) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/RegistryAwareClassLoaderHierarchyHasher.java:38]: The documentation says

> Returns the parent class loader for delegation. Some implementations may use null to represent the bootstrap class loader. This method will return null in such implementations if this class loader's parent is the bootstrap class loader.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Pull request merged**
